### ISSUE TITLE
WindowViewController prefersStatusBarHidden overriding

### ISSUE THREAD

**Issue by giofid** [SwiftKickMobile/SwiftMessages#129]
Hi,
I think that the implementation of `prefersStatusBarHidden` overriding in `WindowViewController` creates problems when  you rotate device with a message showed.
In particular, we'll see a flickering of the status bar if you rotate device with a message showed . This's my current SwiftMessage configuration
```
open class func showWarningMessage(_ msg: String, title: String) {
        let view = MessageView.viewFromNib(layout: .messageView)
        view.configureContent(title: title, body: msg)
        view.configureTheme(.warning, iconStyle: .light)
        view.button?.isHidden = true
        var config = SwiftMessages.defaultConfig
        config.presentationContext = .window(windowLevel: UIWindowLevelStatusBar)
        // Show
        warningMessage.show(config: config, view: view)
    }
```
A fix could be to introduce in `Config` struct a `public var prefersStatusBarHidden: Bool?` property and change `prefersStatusBarHidden` in `WindowViewController` like this
```
override var prefersStatusBarHidden: Bool {
        return config.prefersStatusBarHidden ?? super.prefersStatusBarHidden
    }
```
`statusBarStyle` and `statusBarHidden` in `UIApplication` are deprecated since iOS 9, so I would change `preferredStatusBarStyle` implementation too:
```
override public var preferredStatusBarStyle: UIStatusBarStyle {
        return config.preferredStatusBarStyle ?? super.preferredStatusBarStyle
    }
```
If you want I can submit a pull request.

Thanks,
Giorgio

**Comment by wtmoose** [comment]
Hi. I'm having trouble understanding and reproducing the issue. I can replicate the scenario you've described above using the Demo app, but I don't see any flickering when I rotate my device. Of course, the status bar is covered by the `.window(windowLevel: UIWindowLevelStatusBar)`, so I'm unable to visualize when or how it's flickering.

Can you reproduce the issue with the Demo app (on the Explore screen)? And if not, how can I reproduce it?

Otherwise, your changes sound reasonable. But I'd like to reproduce the problem myself before making any changes.

**Comment by giofid** [comment]
Hi,
thank you for your reply.
Yes, you can reproduce the issue with the Demo app. Below the steps to reproduce it (tested on iPhone 7 (iOS 11) Simulator)

1. Go to _Explore_ section
2. Select _Status bar Level_ as **Presentation Context**
3. Show message
4. With message showed, rotate device in landscape
5. Hide message and you'll see a flickering of the status bar

Of course problem is only present in iPhone device, where status bar is hidden in landscape orientation and is shown in portrait orientation.

**Comment by wtmoose** [comment]
OK got it, thanks. I'm wondering if this is related to #132, where the message view's window is becoming the key window since iOS 11.

**Comment by giofid** [comment]
I just tried on iPhone 7 (10.3.1) Simulator and the issue is also present; so I don't think is related to some changes in iOS 11.

**Comment by wtmoose** [comment]
Could you try the `work/orientation` branch and let me know if it works for you? It should just work out-of-the-box without any extra configuration.

**Comment by giofid** [comment]
Hi Timothy,
I just tried the **Demo** app in `work/orientation` branch and it now works!
But I see some constraints warning in console:
```
2017-11-07 11:11:32.795799+0100 Demo[3298:65406] [LayoutConstraints] Unable to simultaneously satisfy constraints.
	Probably at least one of the constraints in the following list is one you don't want. 
	Try this: 
		(1) look at each constraint and try to figure out which you don't expect; 
		(2) find the code that added the unwanted constraint or constraints and fix it. 
(
    "<NSLayoutConstraint:0x600000284d30 UILayoutGuide:0x6000001a3aa0'TrailingBarGuide(0x7f8ba8e0b3a0)'.height == 32   (active)>",
    "<NSLayoutConstraint:0x600000289fb0 _UIButtonBarStackView:0x7f8ba8c716e0.top == UILayoutGuide:0x6000001a3aa0'TrailingBarGuide(0x7f8ba8e0b3a0)'.top   (active)>",
    "<NSLayoutConstraint:0x60000028a000 _UIButtonBarStackView:0x7f8ba8c716e0.bottom == UILayoutGuide:0x6000001a3aa0'TrailingBarGuide(0x7f8ba8e0b3a0)'.bottom   (active)>",
    "<NSLayoutConstraint:0x60000008eec0 UIButton:0x7f8ba8c4a7b0'SHOW'.height == 33   (active)>",
    "<NSLayoutConstraint:0x60000008ed80 V:|-(0)-[UIButton:0x7f8ba8c4a7b0'SHOW']   (active, names: '|':UIView:0x7f8ba8c239a0 )>",
    "<NSLayoutConstraint:0x60000009ecd0 V:[UIButton:0x7f8ba8c4a7b0'SHOW']-(0)-|   (active, names: '|':UIView:0x7f8ba8c239a0 )>",
    "<NSLayoutConstraint:0x600000289b00 'UISV-canvas-connection' UILayoutGuide:0x6000001a3f00'UIViewLayoutMarginsGuide'.top <= UIView:0x7f8ba8c239a0.top   (active)>",
    "<NSLayoutConstraint:0x600000287300 'UISV-canvas-connection' UILayoutGuide:0x6000001a3f00'UIViewLayoutMarginsGuide'.centerY == UIView:0x7f8ba8c239a0.centerY   (active)>",
    "<NSLayoutConstraint:0x600000289d80 'UIView-bottomMargin-guide-constraint' V:[UILayoutGuide:0x6000001a3f00'UIViewLayoutMarginsGuide']-(0)-|   (active, names: '|':_UIButtonBarStackView:0x7f8ba8c716e0 )>",
    "<NSLayoutConstraint:0x600000289ce0 'UIView-topMargin-guide-constraint' V:|-(0)-[UILayoutGuide:0x6000001a3f00'UIViewLayoutMarginsGuide']   (active, names: '|':_UIButtonBarStackView:0x7f8ba8c716e0 )>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x60000008eec0 UIButton:0x7f8ba8c4a7b0'SHOW'.height == 33   (active)>

Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKit/UIView.h> may also be helpful.
```
Anyway, they are also present in `master` branch, so I don't think are related to this problem.

Lastly, I want to point out again that `statusBarStyle` and `statusBarHidden` in `UIApplication` are deprecated since iOS 9, so I'd change `preferredStatusBarStyle` implementation too:

```
override public var preferredStatusBarStyle: UIStatusBarStyle {
        return config.preferredStatusBarStyle ?? super.preferredStatusBarStyle
    }
```

**Comment by wtmoose** [comment]
Thanks. Updated.

**Comment by wtmoose** [comment]
Fixed in 4.1.0 release
